src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -74,6 +133,59 @@ private class AggregatingTagger : ForegroundThreadAffinitizedObject, IAccurateTa
                 _owner._diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
             }
 
+            private void OnWorkspaceChanged(object sender, EventArgs e)
+            {
+                this.AssertIsForeground();
+
+                // Disconnect from the old workspace notifications and hook up to the new ones.
+                lock (_gate)
+                {
+                    if (_workspace != null)
+                    {
+                        _workspace.DocumentActiveContextChanged -= OnDocumentActiveContextChanged;
+                    }
+
+                    _workspace = _workspaceRegistration.Workspace;
+                    if (_workspace != null)
+                    {
+                        _workspace.DocumentActiveContextChanged += OnDocumentActiveContextChanged;
+                    }
+                }
+
+                // We started tracking another document.  Clear out everything we've stored up so far.
+                ResetCurrentDocumentIdIfNecessary();
+            }
+
+            private void OnDocumentActiveContextChanged(object sender, DocumentActiveContextChangedEventArgs e)
+            {
+                this.AssertIsForeground();
+                this.ResetCurrentDocumentIdIfNecessary();
+            }
+
+            private void ResetCurrentDocumentIdIfNecessary()
+            {
+                this.AssertIsForeground();
+
+                var document = _subjectBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+
+                // Safe to read _currentDocumentId here, we are the fg thread.
```

Comment:
The comment on _gate doesn't state that unlocked foreground thread reads are OK...